xls/modules/rle/rle.x

```diff
@@ -0,0 +1,338 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This file implements a parametric RLE encoder
+//
+// The encoder uses Run Length Encoding (RLE) to compress the input stream of
+// repeating symbols to the output stream that contains the symbols and
+// the number of its consequect occurrences in the input stream.
+// Both the input and the output channels use additional `last` flag
+// that indicates whether the packet ends the transmission. After sending
+// the last packet encoder dumps all the data to the output stream.
+// The behavior of the encoder is presented on the waveform below:
+//                      ──────╥─────╥─────╥─────╥─────╥─────╥─────╥─────╥────
+// next evaluation      XXXXXX║ 0   ║ 1   ║ 2   ║ 3   ║ 4   ║ 5   ║ 6   ║ ...
+//                      ──────╨─────╨─────╨─────╨─────╨─────╨─────╨─────╨────
+//                      ──────╥───────────╥─────╥─────╥─────╥─────╥──────────
+// symbol               XXXXXX║ A         ║ B   ║XXXXX║ B   ║ C   ║XXXXXXXXXX
+// (input channel)      ──────╨───────────╨─────╨─────╨─────╨─────╨──────────
+// last                                   ┌─────┐           ┌─────┐
+// (input channel)      ──────────────────┘     └───────────┘     └──────────
+//                      ╥─────╥─────╥─────╥─────╥─────╥─────╥─────╥──────────
+// state.prev_symbol    ║ 0   ║ A   ║ A   ║ B   ║ 0   ║ B   ║ C   ║ 0
+// (set state value)    ╨─────╨─────╨─────╨─────╨─────╨─────╨─────╨──────────
+//                      ╥─────╥─────╥─────╥─────╥─────╥─────╥─────╥──────────
+// state.prev_count     ║ 0   ║ 1   ║ 2   ║ 1   ║ 0   ║ 1   ║ 1   ║ 0
+// (set state value)    ╨─────╨─────╨─────╨─────╨─────╨─────╨─────╨──────────
+//
+// do_send                                ┌───────────┐     ┌───────────┐
+//                      ──────────────────┘           └─────┘           └────
+//                      ──────────────────╥─────╥─────╥─────╥─────╥─────╥────
+// symbol, count        XXXXXXXXXXXXXXXXXX║ A,2 ║ B,1 ║XXXXX║ B,1 ║ C,1 ║XXXX
+// (output channel)     ──────────────────╨─────╨─────╨─────╨─────╨─────╨────
+// last                                         ┌─────┐           ┌─────┐
+// (output channel)     ────────────────────────┘     └───────────┘     └────
+
+import std
+
+// structure containing input data for an RLE encoder
+pub struct EncInData<SYMBOL_WIDTH: u32> {
+    symbol: bits[SYMBOL_WIDTH], // input symbol to compress
+    last: bool,                 // if this is the last symbol
+}
+
+// structure containing output data from an RLE encoder
+pub struct EncOutData<SYMBOL_WIDTH: u32, COUNT_WIDTH: u32> {
+    symbol: bits[SYMBOL_WIDTH], // compressed symbol
+    count: bits[COUNT_WIDTH],   // symbol count
+    last: bool,                 // if this is the last symbol
+}
+
+// structure to preserve the state of an RLE encoder
+pub struct RLEEncState<SYMBOL_WIDTH: u32, COUNT_WIDTH: u32> {
```

Comment:
nit: I think we can simply name this RLEncoderState (since the `E` of RLE is for encoding).

Reply:
I would rather keep it, but I also see no issues with removing it.
I'll change it in this PR and in RLE decoder PR from `RLEDecState` to `RLDecoderState`.

Reply:
Maybe we could spell it completely? `RunLengthEncoder` and and `RunLengthDecoder`.

I'm also not sure if we need to prefix the type of the state by anything, since those are not public and only used in this file.